src/lib/user_command_input/user_command_input.ml

```diff
@@ -0,0 +1,206 @@
+  module Stable = struct
+    module V1 = struct
+      type t =
+        ( Common.Stable.V1.t
+        , User_command_payload.Body.Stable.V1.t )
+        User_command_payload.Poly.Stable.V1.t
+      [@@deriving sexp, to_yojson]
+
+      let to_latest = Fn.id
+    end
+  end]
+
+  type t = Stable.Latest.t [@@deriving sexp, to_yojson]
+
+  let create ~fee ~nonce_opt ~valid_until ~memo ~body : t =
+    {common= Common.create ~fee ~nonce_opt ~valid_until ~memo; body}
+
+  let to_user_command_payload (t : t) ~inferred_nonce :
+      (User_command_payload.t, string) Result.t =
+    let open Result.Let_syntax in
+    let%map common = Common.to_user_command_common t.common ~inferred_nonce in
+    {User_command_payload.Poly.common; body= t.body}
+end
+
+module Sign_choice = struct
+  [%%versioned
+  module Stable = struct
+    module V1 = struct
+      type t =
+        | Signature of Signature.Stable.V1.t
+        | Hd_index of Unsigned_extended.UInt32.Stable.V1.t
+        | Keypair of Keypair.Stable.V1.t
+      [@@deriving sexp, to_yojson]
+
+      let to_latest = Fn.id
+    end
+  end]
+
+  type t = Stable.Latest.t =
+    | Signature of Signature.t
+    | Hd_index of Unsigned_extended.UInt32.t
+    | Keypair of Keypair.t
+  [@@deriving sexp, to_yojson]
+end
+
+[%%versioned
+module Stable = struct
+  module V1 = struct
+    type t =
+      ( Payload.Stable.V1.t
+      , Public_key.Compressed.Stable.V1.t
+      , Sign_choice.Stable.V1.t )
+      User_command.Poly.Stable.V1.t
+    [@@deriving sexp, to_yojson]
+
+    let to_latest = Fn.id
+  end
+end]
+
+type t = Stable.Latest.t [@@deriving sexp, to_yojson]
+
+let create ?(nonce_opt = None) ~fee ~valid_until ~memo ~body ~sender
+    ~sign_choice () : t =
+  let payload = Payload.create ~fee ~nonce_opt ~valid_until ~memo ~body in
+  {payload; sender; signature= sign_choice}
+
+let sign ~sender ~(user_command_payload : User_command_payload.t) = function
+  | Sign_choice.Signature signature ->
+      Option.value_map
+        ~default:(Deferred.return (Error "Invalid_signature"))
+        (User_command.create_with_signature_checked signature sender
+           user_command_payload)
+        ~f:Deferred.Result.return
+  | Keypair sender_kp ->
+      Deferred.Result.return (User_command.sign sender_kp user_command_payload)
+  | Hd_index hd_index ->
+      Secrets.Hardware_wallets.sign ~hd_index
+        ~public_key:(Public_key.decompress_exn sender)
+        ~user_command_payload
+
+let inferred_nonce ~f ~(sender : Public_key.Compressed.t) ~nonce_map =
+  let update_map = Public_key.Compressed.Map.set nonce_map ~key:sender in
+  match Public_key.Compressed.Map.find nonce_map sender with
+  | Some nonce ->
+      (*multiple user commands from a sender*)
+      let next_nonce = Account_nonce.succ nonce in
+      let updated_map = update_map ~data:next_nonce in
+      Ok (next_nonce, updated_map)
+  | None -> (
+    match Participating_state.active (f sender) |> Option.join with
+    | None ->
+        Error
```

Comment:
Lift this error into  `get_inferred_nonce_from_transaction_pool_and_ledger` in `Coda_lib`. In this context we only know that applying an arbitrary `f` didn't return an `Active (Some_)`.